packages/service-library/src/servicelib/background_task.py

```diff
@@ -1,136 +1,115 @@
 from tenacity.wait import wait_fixed
 
-from .decorators import async_delayed
-from .logging_utils import log_catch, log_context
+from .async_utils import cancel_wait_task, with_delay
+from .logging_utils import log_context
 
 _logger = logging.getLogger(__name__)
 
 
 _DEFAULT_STOP_TIMEOUT_S: Final[int] = 5
-_MAX_TASK_CANCELLATION_ATTEMPTS: Final[int] = 3
-
-
-class PeriodicTaskCancellationError(OsparcErrorMixin, Exception):
-    msg_template: str = "Could not cancel task '{task_name}'"
 
 
 class SleepUsingAsyncioEvent:
-    """Sleep strategy that waits on an event to be set."""
+    """Sleep strategy that waits on an event to be set or sleeps."""
 
     def __init__(self, event: "asyncio.Event") -> None:
         self.event = event
 
-    async def __call__(self, timeout: float | None) -> None:
+    async def __call__(self, delay: float | None) -> None:
         with contextlib.suppress(TimeoutError):
-            await asyncio.wait_for(self.event.wait(), timeout=timeout)
+            await asyncio.wait_for(self.event.wait(), timeout=delay)
             self.event.clear()
 
 
-async def _periodic_scheduled_task(
-    task: Callable[..., Awaitable[None]],
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+def periodic(
     *,
     interval: datetime.timedelta,
-    task_name: str,
-    early_wake_up_event: asyncio.Event | None,
-    **task_kwargs,
-) -> None:
-    # NOTE: This retries forever unless cancelled
-    nap = (
-        asyncio.sleep
-        if early_wake_up_event is None
-        else SleepUsingAsyncioEvent(early_wake_up_event)
-    )
-    async for attempt in AsyncRetrying(
-        sleep=nap,
-        wait=wait_fixed(interval.total_seconds()),
-    ):
-        with attempt:
-            with log_context(
-                _logger,
-                logging.DEBUG,
-                msg=f"iteration {attempt.retry_state.attempt_number} of '{task_name}'",
-            ), log_catch(_logger):
-                await task(**task_kwargs)
+    raise_on_error: bool = False,
+    early_wake_up_event: asyncio.Event | None = None,
+) -> Callable[
+    [Callable[P, Coroutine[Any, Any, None]]], Callable[P, Coroutine[Any, Any, None]]
+]:
+    """Calls the function periodically with a given interval.
+
+    Arguments:
+        interval -- the interval between calls
+
+    Keyword Arguments:
+        raise_on_error -- If False the function will be retried indefinitely unless cancelled.
+                          If True the function will be retried indefinitely unless cancelled
+                          or an exception is raised. (default: {False})
+        early_wake_up_event -- allows to awaken the function before the interval has passed. (default: {None})
+
+    Returns:
+        coroutine that will be called periodically (runs forever)
+    """
 
+    def _decorator(
+        func: Callable[P, Coroutine[Any, Any, None]],
+    ) -> Callable[P, Coroutine[Any, Any, None]]:
+        nap = (
```

Comment:
hahah I like the nap function. both the name and the policy. 

This is a **sleep policy** (i.e. defines "how to sleep") provided as an argument of the decorator. 
Beautiful!